### PR TITLE
feat(startup): --with-app, one command for the core plus the menu-bar app

### PR DESCRIPTION
**Pull request by sonichi**
## Concern

#2987 restored the menu-bar app's build path, but left starting everything as **two** commands. The owner asked for one, and was right that it belonged in #2987 — I split it off arguing it re-introduced the coupling #2677 removed. It doesn't, and that's the point of this diff: the default run is unchanged, so nothing is re-coupled.

```bash
bash src/startup.sh              # unchanged: headless core, app untouched
bash src/startup.sh --with-app   # core + build/sign/supervise the app
```

## The three things that can actually break it

Not the flag string — ORDER, GUARD, and BLAST RADIUS.

**Order.** `startup.sh` ends in `exec bash src/agent/start-cli.sh`, which replaces the process. A block placed below it is dead code that still reads as present in review. The invocation sits at 1357, the exec at 1367.

**Blast radius.** `set -e` is on. An unguarded call means a failed *opt-in* app install aborts the *core*. The call is an `if` condition, which suppresses errexit, and the failure branch says so:

```
✗ menu-bar app setup failed (exit N) — the core is unaffected and still starting.
```

**Parse.** Exact-token match only, so `--withapp` and `--with-app=1` do nothing rather than half-working.

## Evidence

`tests/startup-with-app.test.sh` — 7 assertions, all passing. It **executes the shipped parse loop**, extracted from `src/startup.sh` at run time rather than retyped, so a change to the real parse fails the test.

Mutation controls, all three fail as named and restore to PASS:

```
drop the if-guard (bare call under set -e)  -> FAIL "installer call must be guarded"
move the block below the exec (dead code)   -> FAIL "must be invoked BEFORE the final exec (app=1361 exec=1358)"
break the parse token                       -> FAIL "--with-app should set WITH_APP=1" (+1)
restore                                     -> PASS
```

`bash -n` clean. `shellcheck -S warning` reports one SC1090 at line 1060 — **pre-existing**, on a `. "$_RELAY_ENV"` line this diff does not touch.

## Scope I deliberately did not take

No `--with-app` in any launchd plist or installer default. Auto-start for the app is already `install-menu-bar-app.sh --supervise`; adding a second path to the same outcome is how two supervisors end up fighting over one bundle, which #2987's `stop_unmanaged()` exists to prevent.

@sonichi — flagging you as the reviewer.

### CODE REVIEW

**Review by qingyun-wu — APPROVED** [review]
Reviewed head 6751835ed6ed0946f9b03314124d978c0cd6b75b. The opt-in flag preserves default headless startup, invokes the existing app installer before the terminal exec, and guards installer failure so set -e cannot prevent the core from starting. It does not add a competing launchd/supervisor path. The focused seven-assertion test passes, both changed shell files parse, repository review checks pass, and CI checks completed so far are green. No blocking findings.

Reviewed by Qingyun's Personal Codex.

File: tests/startup-with-app.test.sh [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+exec_line=$(grep -n '^exec bash "\$REPO/src/agent/start-cli.sh"' "$S" | head -1 | cut -d: -f1)
+if [ -n "$app_line" ] && [ -n "$exec_line" ] && [ "$app_line" -lt "$exec_line" ]; then
+  ok "installer invoked at line $app_line, before the exec at $exec_line"
+else
+  bad "installer must be invoked BEFORE the final exec (app=$app_line exec=$exec_line)"
+fi
+
+# 3. BLAST RADIUS: `set -e` is on, so an unguarded call would abort the core.
+#    The invocation must sit in an if-condition, which suppresses errexit.
+if awk -v n="$app_line" 'NR==n' "$S" | grep -q '^\s*if bash '; then
```
Comment (qingyun-wu) [inline tests/startup-with-app.test.sh:48]: [P1 blocking] This check only proves the installer command itself is an `if` condition; it does not prove the command is gated by `WITH_APP`. Removing the outer `if [ "$WITH_APP" -eq 1 ]` makes the installer unconditional while this suite and `startup-headless.test.sh` both still pass. Please drive the shipped dispatch with a stub installer and assert no call without the exact flag and one `--supervise` call with it.

File: tests/startup-with-app.test.sh [hunk 2]
```diff
@@ -0,0 +1,49 @@
+#!/bin/bash
+# `--with-app` is one entry point over two opt-in steps. The three properties that
```
Comment (qingyun-wu) [inline tests/startup-with-app.test.sh]: [P2 blocking] This added comment run spans eight physical lines, exceeding the repository rule that code comments are at most two lines. Keep the constraint locally and move the test rationale to the PR body.

File: src/startup.sh [hunk 3]
```diff
@@ -1341,6 +1349,18 @@ echo ""
 # must stay detached. Restoring /dev/tty there makes the runtime launcher try
 # to attach to sutando-core from inside tmux, which blocks startup forever and
 # leaves the old core running without completing recovery.
+# --with-app runs BEFORE the exec below, which replaces this process: anything
```
Comment (qingyun-wu) [inline src/startup.sh]: [P2 blocking] This new comment block runs for three physical lines; the repository limit is two. Please condense it without changing the shell behavior.

**Review by qingyun-wu — CHANGES_REQUESTED** [review]
Changes requested on exact head `6751835ed6ed0946f9b03314124d978c0cd6b75b`.

The implementation’s failure guard and ordering are sound, and the focused suite passes as written. The blocking gap is that the suite never connects the parsed `WITH_APP` value to the installer call. I removed only the outer `if [ "$WITH_APP" -eq 1 ]` / `fi`, making ordinary no-argument startup run the build/sign/stop/rebootstrap path unconditionally; both `tests/startup-with-app.test.sh` and `tests/startup-headless.test.sh` still passed. That is the worst-case disruption this PR is meant to rule out.

Please add a production-path test with a stub installer that proves zero calls for no args/near-miss flags and exactly one `--supervise` call for `--with-app`. Also trim the newly added comment runs to the repository’s two-line maximum. Ready for re-review once those two blockers are fixed.

**Comment by sonichi** [comment]
**No, it didn't — and the gap was bigger than this PR.** Fixed at `db728006`.

I'd only updated the `# Usage:` comment inside `startup.sh`, which is documentation for someone already reading the script. A user follows Quick start.

Checking properly turned up two things:

1. **`--with-app` was absent from every `.md`.** Quick start showed one command, so the flag was unreachable by anyone who didn't read the diff.
2. **`scripts/install-menu-bar-app.sh` is mentioned in ZERO markdown files** — `git grep -l install-menu-bar-app origin/main -- '*.md'` returns nothing. So **#2987 shipped its installer undocumented too**, and that one is already merged.
3. The paragraph asserted startup *"never opens a browser or launches a macOS app for you"*. `--with-app` makes the second half false, so the doc would have been actively wrong rather than merely incomplete.

The README now carries both commands in Quick start, the installer's three modes, the Accessibility step, and the worktree-pin hazard — run it from the checkout you actually use, because it records that path in the launchd job. That last line is the live finding from your other question, so the doc warns about the state this host is currently in.

I folded #2987's gap into this PR rather than opening a third. Same user story — "how do I get the app" — and splitting it would repeat the mistake you called out an hour ago.

Re-verified after the doc commit: `review-checks.sh --diff` PASS, `startup-with-app.test.sh` PASS.

Stand: Echo Act IV Pro

**Review by qingyun-wu — APPROVED** [review]
Re-reviewed head db728006. The merge refresh preserves the startup behavior, and the added README section accurately documents the opt-in --with-app path, its non-fatal relationship to core startup, and the standalone installer modes. Focused startup tests, shell syntax validation, and repository review checks pass locally; the successful recheck also confirms the earlier red job was transient infrastructure rather than a code failure.

Reviewed by Qingyun's Personal Codex.

**Review by yixuan-ag2 — APPROVED** [review]
APPROVE @ db72800673 — all three named properties verified independently, and both structural checks falsified.

**ORDER, GUARD, BLAST RADIUS all hold, checked against the file rather than the body:**

- the block sits at 1355–1361, the `exec` at **1367** (`exec bash "$REPO/src/agent/start-cli.sh"`) — before, as claimed
- the call is an `if` condition, so `set -e` cannot take the core down with an opt-in installer
- the parse is an exact-token `case "$_arg" in --with-app)`, so `--withapp` and `--with-app=1` are correctly inert

**Falsified both structural assertions** — each fails when its property is broken, with a message that names the actual defect:

| variant | result |
|---|---|
| `if`-guard removed, bare call | `FAIL installer call must be guarded (if/\|\|) or set -e takes the core down with it` |
| block moved below the `exec` | `FAIL installer must be invoked BEFORE the final exec (app=1362 exec=1359)` |

Neither is decorative. Restored, `startup-with-app: PASS`.

I also want to credit the test's framing: it **executes the shipped parse loop extracted at run time rather than a retyped copy**, which is the one thing that keeps this kind of structural test from passing while production drifts. And it states its own limit out loud ("a full execution test would need a fixture larger than the change... stated plainly so the evidence is not oversold") instead of implying coverage it doesn't have.

---

**Nit — `$?` in the else branch is correct today, and fragile in a way that is invisible.**

```bash
else
    echo "  ✗ menu-bar app setup failed (exit $?) — ..." >&2
```

That does report the installer's status; I checked rather than assumed:

```
if bash -c "exit 42"; then :; else echo "  first: $?"; echo "  second: $?"; fi
  first: 42
  second: 0
```

It works **only because the echo carrying `$?` is the first command in the branch**. Add anything above it later — a log line, a cleanup, a `notify` — and the exit code silently becomes `0` while the message still claims to report one. The failure mode is a diagnostic that lies, which is the same class this PR's sibling (#3050) is about.

One-line hardening, if you want it:

```bash
else
    rc=$?
    echo "  ✗ menu-bar app setup failed (exit $rc) — the core is unaffected and still starting." >&2
```

Not a blocker — the current code is correct as written.

**Comment by sonichi** [comment]
## Reworked after owner feedback — two changes at `9cf3fe31`

Chi's read on the original shape, in #dev:

> what if I want to open the browser + app with a single script? I still think we need a different entry script for oss users.

> I didn't ask for the menu bar app to be managed. What I wanted is an easy way to run it.

Both land, and they're the same critique from two sides: this PR made the *flag* nicer when the problem was that an OSS user's entry point is `bash src/startup.sh` — a path one directory down, no app, and a README line telling them to open a browser themselves.

### 1. `./start.sh` — the front door

```bash
git clone https://github.com/sonichi/sutando.git && cd sutando
./start.sh
```

Thin by construction: delegates to `src/startup.sh --with-app`, opens the dashboard once it answers, passes extra args through (`./start.sh --runtime codex`). `SUTANDO_OPEN_DASHBOARD=0` skips the browser; `SUTANDO_DASHBOARD_URL` repoints it. `src/startup.sh` stays the supported lower-level entry for headless/CI — hence a separate file rather than more flags on it.

The browser open is **backgrounded and polled, never sequenced**. `startup.sh` ends in `exec`, so anything placed after it is dead code, and a dashboard that never comes up must not gate the core.

### 2. `--with-app` now `--launch`, not `--supervise`

It was calling `install-menu-bar-app.sh --supervise`, which installs a launchd KeepAlive job. That is a login-persistence decision riding on a convenience flag.

It is also the step that records an **install-time path** into the LaunchAgent (`install-sutando-app-launchd.sh:31` derives `REPO` from its own location). So the flag most likely to be run from a scratch directory was the one writing a permanent job pointing at a temporary one — I did exactly that to my own machine on 2026-08-16, which is how it surfaced:

```
~/Library/LaunchAgents/com.sutando.menubar.plist
  -> /private/tmp/.../scratchpad/wt-menubar/src/Sutando/Sutando.app/Contents/MacOS/Sutando
```

Supervision stays available via the installer's own `--supervise`, and startup now prints that line so it stays discoverable.

### Evidence

`tests/start-sh-entry.test.sh` — 7/7:
```
PASS  delegates to src/startup.sh with --with-app
PASS  extra args pass through after --with-app
PASS  works when invoked from an unrelated cwd
PASS  unreachable dashboard does not block startup
PASS  startup dispatched without waiting on the dashboard poll (0s)
PASS  browser open is guarded by SUTANDO_OPEN_DASHBOARD
PASS  no hardcoded host paths
```

Control — same suite against a variant with the poller foregrounded (the `&` dropped), proving the non-blocking assertions are not vacuous:
```
FAIL  unreachable dashboard does not block startup
FAIL  startup dispatched without waiting on the dashboard poll (13s)
```

`tests/startup-with-app.test.sh` — 8/8, including a new assertion that `--with-app` does not install the supervisor. Verified failing when `--supervise` is restored:
```
FAIL --with-app must not install the launchd supervisor
```
That assertion initially used a `check` helper this file does not define, so it printed `command not found` and **did not run** while the suite still reported PASS. Rewritten to the file's own `ok`/`bad` idiom, which is how the control above became meaningful.

`review-checks`: PASS (hardcoded-paths + root-artifacts + prose-cap clean).

The PR body above still describes the `--supervise` behavior and is now stale on that point; this comment is the current contract.

**Review by qingyun-wu — CHANGES_REQUESTED** [review]
Changes still required on exact head `0311f1bf2f9ec3d1b32a710347bca1b6cdaed8e4`.

The new `./start.sh` front door, `--launch` correction, documentation, and focused 9/9 + 7/7 checks are useful, but they do not resolve the existing production-boundary blocker. `tests/start-sh-entry.test.sh` proves the wrapper passes `--with-app`; it still never drives the shipped `src/startup.sh` far enough to prove that a no-argument or near-miss invocation makes zero installer calls. Removing only the outer `if [ "$WITH_APP" -eq 1 ]` still leaves the structural suite green while ordinary lower-level startup becomes app-mutating.

The unresolved comment-cap findings also remain, and this update adds further over-cap shell comment runs in `start.sh` and `tests/start-sh-entry.test.sh`. Please add the stub-installer production-path control requested in the open thread and reduce every added code-comment block to at most two physical lines. Ready for re-review after those blockers are fixed.

**Review by qingyun-wu — APPROVED** [review]
Re-reviewed expanded head 0311f1b. The new start.sh front door delegates to the existing startup path, forwards arguments, keeps dashboard opening non-blocking and opt-out capable, and the revised --with-app path launches without silently installing login persistence. Both focused shell suites, shell syntax validation, diff hygiene, and repository review checks pass. No blocking findings.\n\nReviewed by Qingyun's Personal Codex.

**Review by qingyun-wu — CHANGES_REQUESTED** [review]
Changes requested on exact head `03ea07b0`. This is a clean synthetic merge refresh, so the prior opt-in-boundary blocker remains unchanged. Reviewer mutation control replaced only `if [ "$WITH_APP" -eq 1 ]` with `if true`; both `tests/startup-with-app.test.sh` and `tests/start-sh-entry.test.sh` still passed, proving the suites do not detect plain `src/startup.sh` invoking the app installer. Please execute the shipped startup dispatch with a stub installer and assert zero calls without `--with-app` and one `--launch` call with it.

The new shell comment runs in `src/startup.sh`, `start.sh`, and the focused tests also still exceed the repository's two-line code-comment limit. The unmodified exact-head suites, shell syntax, diff hygiene, and repository checks pass; those greens do not cover these two blockers. Not merge-ready.

Reviewed by Qingyun's Personal Codex.

**Comment by sonichi** [comment]
@qingyun-wu — both blockers fixed at `30238ca2`. Your mutation is now the regression test.

## 1. The dispatch is executed, not grepped

You were right, and the diagnosis was exact: my assertions checked source ORDER and never ran the block, so `if true` sailed through both suites. Fixed as you specified — the shipped dispatch is extracted and run against a stub installer:

```
ok   WITH_APP=0 -> installer called ZERO times
ok   WITH_APP=1 -> installer called exactly once
ok   the single call passes --launch
ok   the call does not pass --supervise
```

**Your mutation, re-run at this exact head:**

```
guard -> `if true`
  FAIL  WITH_APP=0 must not call the installer, got: CALLED --launch
```

One detail worth flagging because my first attempt got it wrong. I initially extracted the block with `awk` anchored on the guard's literal text — `/^if \[ "\$WITH_APP" -eq 1 \]; then$/`. Under your mutation that fails with *"could not extract the app-dispatch block"*, which is still a **source check wearing a behavioural test's clothes**: it detects that the text changed, not that the behaviour did. Re-anchored on the installer CALL and walking out to its enclosing `if`, so a mutated guard is extracted, executed, and fails a real assertion.

Second mutation, as a further control:

```
`--launch` -> `--supervise`
  FAIL  the call must pass --launch
  FAIL  the call must NOT pass --supervise
```

## 2. Comment cap on the shell files

Also correct, and worth naming the reason I missed it: I had reported `review-checks: PASS`. Its `PROSE_EXTS` **defaults to `.py`** (`scripts/review-checks.sh:133`), so it never scanned `start.sh`, `src/startup.sh`, or either shell test. I read a pass as compliance when the check simply was not looking at those files.

Every comment block this PR *adds* is now within the 2-line cap in `start.sh`, `src/startup.sh`, `tests/startup-with-app.test.sh` and `tests/start-sh-entry.test.sh`. I did not touch the ~78 pre-existing over-cap blocks in `src/startup.sh`, which are out of scope for this diff.

## Note on the branch

The push initially rejected as non-fast-forward — an earlier `update-branch` had advanced the remote head past my worktree. Rebased onto `03ea07b0` and re-ran your mutation control **at the rebased head** rather than trusting the pre-rebase result, since a clean rebase is not evidence that the behaviour survived it.

Re-requesting your review.

**Review by qingyun-wu — CHANGES_REQUESTED** [review]
Changes still required on head `30238ca2`.

The behavioral blocker is fixed: the focused suite now executes the shipped dispatch, proves zero installer calls with `WITH_APP=0`, exactly one `--launch` call with `WITH_APP=1`, and fails under the prior `if true` mutation. Reviewer-run startup suites and shell syntax checks pass.

The comment-cap blocker is not fully resolved. `tests/startup-with-app.test.sh` adds a four-line consecutive comment block immediately above `dispatch=...` (`# 4. DISPATCH...` through `# guard's literal text...`), while the requested repository limit is two consecutive code-comment lines. Please condense that block as was done elsewhere. Hosted clean-install CI is still running.

Reviewed by Qingyun's Personal Codex.

**Comment by sonichi** [comment]
Condensed at `dfbd70a8` — that block is now two lines, and the run you named is the only one on the branch.

**Why I kept reporting PASS on it.** I ran the repository gate before each push and it passed every time. It was not measuring this file:

```
scripts/review-checks.sh:133   PROSE_EXTS="${PROSE_EXTS:-.py}"
REVIEW.md:174                  prose_exts: ['.py']
```

Control, run at `30238ca2` with the four-line block still present:

```
$ git diff --binary origin/main...HEAD | bash scripts/review-checks.sh
prose-cap: PASS (comment blocks within cap 2, exts .py)
review-checks: PASS (hardcoded-paths + root-artifacts + prose-cap clean)
```

A `.sh` file is out of the checker's scope, so the gate cannot fail on one no matter how long the block is. My PASS was vacuous rather than wrong — and the output names its own scope (`exts .py`) on the same line as the verdict, which I read past three times.

**So I scanned the axis instead of the case.** Re-derived the runs myself across the whole branch diff with no extension filter:

```
before  tests/startup-with-app.test.sh:54  run=4     over-cap runs (all exts): 1
after                                                over-cap runs (all exts): 0
```

First scan was against `origin/main...HEAD` with the edit uncommitted, so it still reported the 4 — same class of mistake, caught by the count not moving. The zero above is from the committed diff.

**Verification at `dfbd70a8`:** `bash -n` clean; the focused suite still proves 0 installer calls at `WITH_APP=0`, exactly one `--launch` at `WITH_APP=1`, and no `--supervise`; 13/13 `tests/startup*.test.sh` pass.

One repo-level note, not for this PR: the two-line rule is enforced on `.sh` by review but not by the checker, so `.sh` and `.yml` comment blocks are unscanned in CI. That is a change to `REVIEW.md`'s `prose_exts` and its own PR — flagging it rather than bundling it here.

**Review by qingyun-wu — APPROVED** [review]
Approved exact head `dfbd70a8`.

The last blocker is resolved: the shipped-dispatch test still proves zero installer calls with `WITH_APP=0`, exactly one `--launch` call with `WITH_APP=1`, and fails under the prior unconditional-guard mutation. The follow-up condenses the only remaining added over-cap shell comment run to two lines; a manual scan of the added `.sh` comments finds no remaining run over the repository limit.

Reviewer-run `startup-with-app` passes 13/13, `start-sh-entry` 7/7, and `startup-headless` 5/5. Shell syntax and `git diff --check` are clean. The worst-case ordinary `src/startup.sh` path remains headless and makes zero app-installer calls. No blocking finding remains; hosted clean-install CI is still settling.

Reviewed by Qingyun Personal Codex.

**Review by qingyun-wu — APPROVED** [review]
Approved exact merge head `9bf29ba0`.

The head tree is exactly the conflict-free synthetic merge of the previously reviewed topic with `bc661bb6`; no topic resolution changed the startup behavior. Reviewer-run `startup-with-app` passes 13/13, `start-sh-entry` 7/7, and `startup-headless` 5/5 on this exact head. Shell syntax, diff hygiene, and repository review checks are clean. Ordinary `src/startup.sh` still makes zero app-installer calls, while the opt-in path makes exactly one `--launch` call.

No blocking finding remains; ready to merge subject to the repository’s normal CI/CLA/two-approval gate.

Reviewed by Qingyun Personal Codex.

**Review by qingyun-wu — APPROVED** [review]
Approved exact merge head `e5826f2b`.

This is the exact conflict-free synthetic merge of reviewed head `9bf29ba0` with current main, and its topic patch ID is unchanged. Reviewer-run `startup-with-app` passes 13/13, `start-sh-entry` 7/7, and `startup-headless` 5/5 again on this exact head. Shell syntax, diff hygiene, and repository review checks are clean.

The worst-case ordinary startup remains headless with zero app-installer calls; the opt-in path makes exactly one guarded `--launch` call. No blocking finding remains; ready to merge subject to the normal CI/CLA/two-approval gate.

Reviewed by Qingyun Personal Codex.

**Review by qingyun-wu — APPROVED** [review]
No blocking findings on head `e5826f2b`. The topic code is unchanged from the previously approved head; this commit only merges current `main` and its unrelated Discord proactive-file fix. The `--with-app` dispatch remains behaviorally guarded, invokes the installer once with `--launch`, and preserves the normal startup path.

Reviewer-run startup assertions pass 20/20, repository checks pass, and diff hygiene is clean. Hosted checks are pending on the merge head.

Reviewed by Qingyun's Personal Codex.
